### PR TITLE
Float16 + 12 block residual network

### PR DESCRIPTION
**Pull request by Akababa**
After much fudging with epsilon constants I believe I've achieved numerical stability on half-precision which has given me a much-needed performance boost. I'm currently training a 12 block network from scratch (results of transfer learning were disappointing) on 22k games where the rating difference between the players was >=600.

### CODE REVIEW

**Comment by RavnaBergsndot** [comment]
The author of leela zero said he had problem making the value part of the network converge with fp16. Could you share your experience with us?

**Comment by Akababa** [comment]
My 12-block network is still not as good as my 7-block one, but I initially had some issues with numerical stability which I (naively) solved by doing softmax outside the network and increasing epsilon constants. I imagine a similar thing can be done with the value head, or using a sigmoid with lower constant. Did he mention if numerical stability was the problem, or something else?

**Comment by RavnaBergsndot** [comment]
No, he's vague about the details. I think I'll study your code first.

**Comment by Akababa** [comment]
Thanks, please do let me know if you find any bugs or insights - for some reason I've been unable to match the performance of my 7-block model.

**Comment by RavnaBergsndot** [comment]
It could be because of overfitting. 12-block is larger than 7-block so it might need more data.

By the way, in Go AIs trained from human games, any AI needs more than 100k games to get any semblance of strength. I don't know if it's the same case in chess but I suggest more data, or less filters (try 64 or 128 first), or both.

**Comment by Akababa** [comment]
Thanks for the input. I'm impatient (and without a good GPU) so my strategy has been to cut regularization tenfold and only run 1-2 epochs per dataset. Is this policy+value+MCTS setup similar to stuff that has been used before?

**Comment by RavnaBergsndot** [comment]
Yes, this is similar to the AGZ setting. They used 4.0 instead of your 1.5 as the value of c_puct though. I think a higher c_puct is important especially when the number of playouts is relatively low.

**Comment by Akababa** [comment]
Thanks, do you have a reference for that? I was looking for their c_puct constant but couldn't find it online.

I think of c_puct as how much you trust your policy (relative to value). If it's 0 the agent only considers the child state with largest expected value (not sure I agree with the 0 initialization for that; perhaps a better choice would be expected value of the current state), and as c->infty it's just picking the policy move.

**Comment by RavnaBergsndot** [comment]
I think my memory was wrong. Here's the value in the original alphago paper and it's 5:

Extended Data Table 5 | Parameters used by AlphaGo

Symbol Parameter Value

β Softmax temperature 0.67

λ Mixing parameter 0.5

nvl Virtual loss 3

nthr Expansion threshold 40

cpuct Exploration constant 5

The agz paper didn't mention the exact value but it said it roughly follows the alphago paper. ("The selection phase is almost identical to AlphaGo Fan")

In my opinion it also depends on the playout number. If the playout number is higher the tree gets expanded deeper, therefore the value estimation would be more accurate (closer to the end of the game). If the playout number is low it's better to trust the policy suggestion more.

**Comment by Akababa** [comment]
Thanks. Interestingly, I found [leela-zero uses 0.85](https://github.com/gcp/leela-zero/blob/master/src/GTP.cpp#L75).

**Comment by RavnaBergsndot** [comment]
Yeah. I also posted something about tuning these hyperparameters on the leelaz project. I guess it's always a good thing to do more experiments but we are all limited by our hardware...

**Comment by Akababa** [comment]
So it turns out my float16 implementation isn't as numerically stable as I thought. Instead of wasting another 2 weeks trying to train this noisy model I'm going back to the original 7 block with float32s. I think it still has quite a bit more capacity left (although I doubt it can encode much long-range tactics)
